Code/GraphMol/ChemReactions/Enumerate/EvenSamplePairs.h

```diff
@@ -0,0 +1,188 @@
+      number of samples, this method performs progressively worse as the
+      number of samples gets larger.
+
+      See EnumeartionStrategyBase for more details.
+*/
+
+class EvenSamplePairsStrategy : public EnumerationStrategyBase {
+  boost::uint64_t m_numPermutationsProcessed;
+
+  std::vector<int64_t> used_count;
+  std::vector<std::vector<size_t> > var_used;
+  std::vector<std::vector<size_t> > pair_used;
+  std::vector<std::vector<size_t> > pair_counts;
+  std::set<size_t> selected;
+
+  size_t seed;     // last seed for permutation (starts at 0)
+  size_t M, a, b;  // random number stuff
+  size_t nslack, min_nslack;
+  size_t rejected_period, rejected_unique;
+  size_t rejected_slack_condition, rejected_bb_sampling_condition;
+
+ public:
+  EvenSamplePairsStrategy()
+      : EnumerationStrategyBase(),
+        m_numPermutationsProcessed(),
+        used_count(),
+        var_used(),
+        pair_used(),
+        pair_counts(),
+        selected(),
+        seed(),
+        M(),
+        a(),
+        b(),
+        nslack(),
+        min_nslack(),
+        rejected_period(),
+        rejected_unique(),
+        rejected_slack_condition(),
+        rejected_bb_sampling_condition() {}
+
+  EvenSamplePairsStrategy(const EvenSamplePairsStrategy &rhs)
+      : EnumerationStrategyBase(rhs),
+        m_numPermutationsProcessed(rhs.m_numPermutationsProcessed),
+        used_count(rhs.used_count),
+        var_used(rhs.var_used),
+        pair_used(rhs.pair_used),
+        pair_counts(rhs.pair_counts),
+        selected(rhs.selected),
+        seed(rhs.seed),
+        M(rhs.M),
+        a(rhs.a),
+        b(rhs.b),
+        nslack(rhs.nslack),
+        min_nslack(rhs.min_nslack),
+        rejected_period(rhs.rejected_period),
+        rejected_unique(rhs.rejected_unique),
+        rejected_slack_condition(rhs.rejected_slack_condition),
+        rejected_bb_sampling_condition(rhs.rejected_bb_sampling_condition) {}
+
+  virtual const char *type() const { return "EvenSamplePairsStrategy"; }
+
+  //! This is a class for enumerating RGroups using Cartesian Products of
+  //! reagents.
+  /*!
+    basic usage:
+
+    \verbatim
+    std::vector<MOL_SPTR_VECT> bbs;
+    bbs.push_back( bbs_for_reactants_1 );
+    bbs.push_back( bbs_for_reactants_2 );
+
+    EvenSamplePairsStrategy rgroups;
+    rgroups.initialize(rxn, bbs);
+    for(size_t i=0; i<num_samples && rgroups; ++i) {
+      MOL_SPTR_VECT rvect = getReactantsFromRGroups(bbs, rgroups.next());
+      std::vector<MOL_SPTR_VECT> lprops = rxn.RunReactants(rvect);
+      ...
+    }
+    \endverbatim
+  */
+  using EnumerationStrategyBase::initialize;
+
+  virtual void initializeStrategy(const ChemicalReaction &, const EnumerationTypes::BBS &);
+
+  //! The current permutation {r1, r2, ...}
+  const EnumerationTypes::RGROUPS &next();
+
+  size_t getPermutationIdx() const { return m_numPermutationsProcessed; }
+
+  operator bool() const { return true; }
```

Comment:
I believe this should be virtual so that it can be over-ridden in derived classes